<a href="https://colab.research.google.com/github/JavokhirJambulov/AIapplicationSystem/blob/main/12204567_Homework_Week_5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Activity 1:

In [2]:
# Import TensorFlow and other libraries
import tensorflow as tf
import numpy as np
import matplotlib.pyplot as plt
# Load the MNIST dataset
mnist = tf.keras.datasets.mnist
# Split the dataset into training and test sets
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()
# Normalize the images
train_images = train_images / 255.0
test_images = test_images/255.0

11490434/11490434 [==============================] - 0s 0us/step


In [5]:
# Function to create and evaluate model with different activation functions
def evaluate_activation (activation_function):
  model = tf.keras. Sequential ([
    tf.keras.layers.Flatten (input_shape=(28, 28)),
    tf.keras.layers.Dense (128, activation=activation_function),
    tf.keras.layers.Dense (10, activation='softmax')
  ])
  model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
  model.fit(train_images, train_labels, epochs=5)
  test_loss, test_acc = model.evaluate(test_images, test_labels)
  print(f"Test accuracy with {activation_function}: {test_acc}")
# Evaluate models with different activation functions
for activation in ['sigmoid', 'tanh', 'relu']:
  evaluate_activation (activation)

Epoch 1/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.3951 - accuracy: 0.8980
Epoch 2/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.1912 - accuracy: 0.9452
Epoch 3/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.1401 - accuracy: 0.9595
Epoch 4/5
1875/1875 [==============================] - 7s 3ms/step - loss: 0.1091 - accuracy: 0.9689
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 0.1030 - accuracy: 0.9682
Test accuracy with sigmoid: 0.9682000279426575
Epoch 1/5
1875/1875 [==============================] - 7s 3ms/step - loss: 0.2814 - accuracy: 0.9186
Epoch 2/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.1368 - accuracy: 0.9602
Epoch 3/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0932 - accuracy: 0.9729
Epoch 4/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0681 - accuracy: 0.9798
Epoch 5/5
313/313 [===========================


Activity 1:

*  Accuracy with sigmoid: 0.9682000279426575

*  Accuracy with tanh:    0.9775000214576721

*  Acuracy with relu:     0.9733999967575073



Activity 2:

In [7]:
# Evaluate models with different learning rates
for Ir in [0.001, 0.01, 0.1]:
  optimizer = tf.keras.optimizers.Adam(learning_rate=Ir)
  model = tf.keras. Sequential ([
  tf.keras.layers. Flatten (input_shape=(28, 28)),
  tf.keras. layers. Dense (128, activation= 'relu'),
  tf.keras. layers. Dense (10, activation='softmax')
  ])
  model.compile(optimizer=optimizer, loss='sparse_categorical_crossentropy', metrics=['accuracy'])
  model.fit(train_images, train_labels, epochs=5, batch_size=32)
  test_loss, test_acc = model.evaluate(test_images, test_labels)
  print(f"Test accuracy with learning rate {Ir}: {test_acc}")

Epoch 1/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.2573 - accuracy: 0.9267
Epoch 2/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.1125 - accuracy: 0.9664
Epoch 3/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0777 - accuracy: 0.9766
Epoch 4/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0585 - accuracy: 0.9820
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 0.0720 - accuracy: 0.9785
Test accuracy with learning rate 0.001: 0.9785000085830688
Epoch 1/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.2430 - accuracy: 0.9283
Epoch 2/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.1553 - accuracy: 0.9564
Epoch 3/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.1447 - accuracy: 0.9615
Epoch 4/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.1288 - accuracy: 0.9650
Epoch 5/5
313/313 [===============

In [8]:
# Evaluate models with different learning rates
for Ir in [0.001, 0.01, 0.1]:
  optimizer = tf.keras.optimizers.Adam(learning_rate=Ir)
  model = tf.keras. Sequential ([
  tf.keras.layers. Flatten (input_shape=(28, 28)),
  tf.keras. layers. Dense (128, activation= 'relu'),
  tf.keras. layers. Dense (10, activation='softmax')
  ])
  model.compile(optimizer=optimizer, loss='sparse_categorical_crossentropy', metrics=['accuracy'])
  model.fit(train_images, train_labels, epochs=5, batch_size=64)
  test_loss, test_acc = model.evaluate(test_images, test_labels)
  print(f"Test accuracy with learning rate {Ir}: {test_acc}")

Epoch 1/5
938/938 [==============================] - 3s 3ms/step - loss: 0.2939 - accuracy: 0.9173
Epoch 2/5
938/938 [==============================] - 5s 5ms/step - loss: 0.1319 - accuracy: 0.9617
Epoch 3/5
938/938 [==============================] - 3s 3ms/step - loss: 0.0934 - accuracy: 0.9730
Epoch 4/5
938/938 [==============================] - 4s 4ms/step - loss: 0.0722 - accuracy: 0.9786
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 0.0779 - accuracy: 0.9774
Test accuracy with learning rate 0.001: 0.977400004863739
Epoch 1/5
938/938 [==============================] - 3s 3ms/step - loss: 0.2181 - accuracy: 0.9341
Epoch 2/5
938/938 [==============================] - 3s 3ms/step - loss: 0.1298 - accuracy: 0.9628
Epoch 3/5
938/938 [==============================] - 4s 4ms/step - loss: 0.1125 - accuracy: 0.9674
Epoch 4/5
938/938 [==============================] - 3s 3ms/step - loss: 0.1028 - accuracy: 0.9711
Epoch 5/5
313/313 [==============================] 

In [9]:
# Evaluate models with different learning rates
for Ir in [0.001, 0.01, 0.1]:
  optimizer = tf.keras.optimizers.Adam(learning_rate=Ir)
  model = tf.keras. Sequential ([
  tf.keras.layers. Flatten (input_shape=(28, 28)),
  tf.keras. layers. Dense (128, activation= 'relu'),
  tf.keras. layers. Dense (10, activation='softmax')
  ])
  model.compile(optimizer=optimizer, loss='sparse_categorical_crossentropy', metrics=['accuracy'])
  model.fit(train_images, train_labels, epochs=5, batch_size=128)
  test_loss, test_acc = model.evaluate(test_images, test_labels)
  print(f"Test accuracy with learning rate {Ir}: {test_acc}")

Epoch 1/5
469/469 [==============================] - 3s 5ms/step - loss: 0.3566 - accuracy: 0.9019
Epoch 2/5
469/469 [==============================] - 2s 4ms/step - loss: 0.1632 - accuracy: 0.9537
Epoch 3/5
469/469 [==============================] - 2s 4ms/step - loss: 0.1175 - accuracy: 0.9667
Epoch 4/5
469/469 [==============================] - 2s 4ms/step - loss: 0.0910 - accuracy: 0.9742
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 0.0872 - accuracy: 0.9737
Test accuracy with learning rate 0.001: 0.9736999869346619
Epoch 1/5
469/469 [==============================] - 3s 5ms/step - loss: 0.2196 - accuracy: 0.9330
Epoch 2/5
469/469 [==============================] - 2s 3ms/step - loss: 0.1080 - accuracy: 0.9671
Epoch 3/5
469/469 [==============================] - 2s 4ms/step - loss: 0.0877 - accuracy: 0.9735
Epoch 4/5
469/469 [==============================] - 2s 4ms/step - loss: 0.0745 - accuracy: 0.9776
Epoch 5/5
313/313 [==============================]

Activity 2:

Batch size: 32 and different learning rates(0.001, 0.01, 0.1)

Test accuracy with learning rate 0.001: 0.9785000085830688

Test accuracy with learning rate 0.01: 0.9656999707221985

Test accuracy with learning rate 0.1: 0.5285000205039978

-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-

Batch size: 64 and different learning rates(0.001, 0.01, 0.1)

Test accuracy with learning rate 0.001: 0.977400004863739

Test accuracy with learning rate 0.01: 0.9552000164985657

Test accuracy with learning rate 0.1: 0.7063999772071838


-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-

Batch size: 128 and different learning rates(0.001, 0.01, 0.1)

Test accuracy with learning rate 0.001: 0.9736999869346619

Test accuracy with learning rate 0.01: 0.9642999768257141

Test accuracy with learning rate 0.1: 0.8715999722480774

Activity 3:

In [11]:
import tensorflow as tf
import numpy as np
import matplotlib.pyplot as plt

# Load the MNIST dataset
mnist = tf.keras.datasets.mnist

# Split the dataset into training and test sets
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

# Normalize the images
train_images = train_images / 255.0
test_images = test_images / 255.0

# Function to create and evaluate model with different activation functions and dropout
def evaluate_activation_with_dropout(activation_function, dropout_rate=0.2):
    model = tf.keras.Sequential([
        tf.keras.layers.Flatten(input_shape=(28, 28)),
        tf.keras.layers.Dense(128, activation=activation_function),
        tf.keras.layers.Dropout(rate=dropout_rate),  # Add a dropout layer
        tf.keras.layers.Dense(10, activation='softmax')
    ])
    model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
    model.fit(train_images, train_labels, epochs=5)
    test_loss, test_acc = model.evaluate(test_images, test_labels)
    print(f"Test accuracy with {activation_function} and dropout: {test_acc}")

# Evaluate models with different activation functions and dropout rates
for activation in ['sigmoid', 'tanh', 'relu']:
    evaluate_activation_with_dropout(activation)


Epoch 1/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.4432 - accuracy: 0.8785
Epoch 2/5
1875/1875 [==============================] - 10s 5ms/step - loss: 0.2308 - accuracy: 0.9329
Epoch 3/5
1875/1875 [==============================] - 13s 7ms/step - loss: 0.1747 - accuracy: 0.9487
Epoch 4/5
1875/1875 [==============================] - 8s 4ms/step - loss: 0.1409 - accuracy: 0.9582
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 0.1016 - accuracy: 0.9707
Test accuracy with sigmoid and dropout: 0.9707000255584717
Epoch 1/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.3245 - accuracy: 0.9062
Epoch 2/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.1858 - accuracy: 0.9453
Epoch 3/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.1444 - accuracy: 0.9554
Epoch 4/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.1224 - accuracy: 0.9618
Epoch 5/5
313/313 [=============

In [12]:
import tensorflow as tf
import numpy as np
import matplotlib.pyplot as plt

# Load the MNIST dataset
mnist = tf.keras.datasets.mnist

# Split the dataset into training and test sets
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

# Normalize the images
train_images = train_images / 255.0
test_images = test_images / 255.0

# Function to create and evaluate regularized model with different activation functions
def evaluate_activation_with_regularization(activation_function, l2_regularization=0.01):
    model = tf.keras.Sequential([
        tf.keras.layers.Flatten(input_shape=(28, 28)),
        tf.keras.layers.Dense(128, activation=activation_function,
                              kernel_regularizer=tf.keras.regularizers.l2(l2_regularization)),  # Add L2 regularization
        tf.keras.layers.Dense(10, activation='softmax')
    ])
    model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
    model.fit(train_images, train_labels, epochs=5)
    test_loss, test_acc = model.evaluate(test_images, test_labels)
    print(f"Test accuracy with {activation_function} and L2 regularization: {test_acc}")

# Evaluate models with different activation functions and L2 regularization strengths
for activation in ['sigmoid', 'tanh', 'relu']:
    evaluate_activation_with_regularization(activation, l2_regularization=0.01)


Epoch 1/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.9792 - accuracy: 0.8538
Epoch 2/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.6699 - accuracy: 0.8809
Epoch 3/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.6088 - accuracy: 0.8827
Epoch 4/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.5764 - accuracy: 0.8853
Epoch 5/5
313/313 [==============================] - 1s 2ms/step - loss: 0.5291 - accuracy: 0.8959
Test accuracy with sigmoid and L2 regularization: 0.8959000110626221
Epoch 1/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.6327 - accuracy: 0.8837
Epoch 2/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.4679 - accuracy: 0.9015
Epoch 3/5
1875/1875 [==============================] - 5s 3ms/step - loss: 0.4421 - accuracy: 0.9083
Epoch 4/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.4182 - accuracy: 0.9165
Epoch 5/5
313/313 [=====

Activity 3:

Baseline model:

Accuracy with sigmoid: 0.9682000279426575

Accuracy with tanh: 0.9775000214576721

Acuracy with relu: 0.9733999967575073

-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-

Model with dropout layer:

Test accuracy with sigmoid and dropout: 0.9707000255584717

Test accuracy with tanh and dropout: 0.972599983215332

Test accuracy with relu and dropout: 0.9779000282287598

You can add a dropout layer by including tf.keras.layers.Dropout in your model definition. The code that was used added a dropout layer with a dropout rate of 0.2 to the model.

-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-

Model with L2 regularization:

Test accuracy with sigmoid and L2 regularization: 0.8959000110626221

Test accuracy with tanh and L2 regularization: 0.9223999977111816

Test accuracy with relu and L2 regularization: 0.9531999826431274

You can add L2 regularization to dense layers using the kernel_regularizer argument in the tf.keras.layers.Dense layer. The code that was used added a L2 regularization with a strength of 0.01 to the dense layers.

Activity 4:

In [13]:
import tensorflow as tf
from sklearn.metrics import precision_score, recall_score, f1_score

# Assuming you have a model that has already been trained
# Make predictions on the test dataset
predicted_labels = model.predict(test_images)

# Assuming that you have ground truth labels for the test dataset
# Convert predicted probabilities to class labels
predicted_classes = np.argmax(predicted_labels, axis=1)

# Calculate precision, recall, and F1-score using TensorFlow
precision = tf.keras.metrics.Precision()(test_labels, predicted_classes)
recall = tf.keras.metrics.Recall()(test_labels, predicted_classes)
f1 = 2 * (precision * recall) / (precision + recall)

print(f"Precision: {precision.numpy()}")
print(f"Recall: {recall.numpy()}")
print(f"F1-score: {f1.numpy()}")


313/313 [==============================] - 1s 3ms/step
Precision: 0.9934864044189453
Recall: 0.9976718425750732
F1-score: 0.9955747127532959


Activity 4:

Precision: 0.9934864044189453
Precision is the metric to use when minimizing false positives is crucial.

Recall: 0.9976718425750732
Recall is the metric to use when minimizing false negatives is critical.

F1-score: 0.9955747127532959
The F1-score is the harmonic mean of precision and recall.

Activity 5:

In [14]:
import tensorflow as tf

# Load the MNIST dataset
mnist = tf.keras.datasets.mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

# Normalize the images
train_images = train_images / 255.0
test_images = test_images / 255.0

# Build the baseline feedforward neural network model
baseline_model = tf.keras.Sequential([
    tf.keras.layers.Flatten(input_shape=(28, 28)),  # Flatten the 28x28 input images
    tf.keras.layers.Dense(128, activation='relu'),  # Fully connected layer with ReLU activation
    tf.keras.layers.Dense(64, activation='relu'),   # Additional fully connected layer
    tf.keras.layers.Dense(10, activation='softmax') # Output layer with 10 classes (digits)
])

# Compile the model
baseline_model.compile(optimizer='adam',
                       loss='sparse_categorical_crossentropy',
                       metrics=['accuracy'])

# Train the model
baseline_model.fit(train_images, train_labels, epochs=5, verbose=2)

# Evaluate the model on the test dataset
test_loss, test_accuracy = baseline_model.evaluate(test_images, test_labels)
print(f"Test accuracy of baseline model: {test_accuracy}")


Epoch 1/5
1875/1875 - 6s - loss: 0.2375 - accuracy: 0.9295 - 6s/epoch - 3ms/step
Epoch 2/5
1875/1875 - 6s - loss: 0.0997 - accuracy: 0.9696 - 6s/epoch - 3ms/step
Epoch 3/5
1875/1875 - 5s - loss: 0.0708 - accuracy: 0.9774 - 5s/epoch - 3ms/step
Epoch 4/5
1875/1875 - 7s - loss: 0.0535 - accuracy: 0.9830 - 7s/epoch - 4ms/step
Epoch 5/5
1875/1875 - 5s - loss: 0.0414 - accuracy: 0.9861 - 5s/epoch - 3ms/step
313/313 [==============================] - 1s 2ms/step - loss: 0.0830 - accuracy: 0.9757
Test accuracy of baseline model: 0.9757000207901001


In [15]:
import tensorflow as tf
from tensorflow.keras import layers, models
mnist = tf.keras.datasets.mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()
train_images, test_images = train_images / 255.0, test_images / 255.0
model_cnn = models.Sequential([
    layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)),
    layers.MaxPooling2D((2, 2)),
    layers.Flatten(),
    layers.Dense(128, activation='relu'),
    layers.Dense(10, activation='softmax')
])

model_cnn.compile(optimizer='adam',
                  loss='sparse_categorical_crossentropy',
                  metrics=['accuracy'])
train_images = train_images.reshape(train_images.shape[0], 28, 28, 1)
test_images = test_images.reshape(test_images.shape[0], 28, 28, 1)
model_cnn.fit(train_images, train_labels, epochs=5)
test_loss, test_acc = model_cnn.evaluate(test_images, test_labels)
print(f"Test accuracy of CNN model: {test_acc}")


Epoch 1/5
1875/1875 [==============================] - 35s 18ms/step - loss: 0.1511 - accuracy: 0.9556
Epoch 2/5
1875/1875 [==============================] - 34s 18ms/step - loss: 0.0531 - accuracy: 0.9834
Epoch 3/5
1875/1875 [==============================] - 34s 18ms/step - loss: 0.0341 - accuracy: 0.9894
Epoch 4/5
1875/1875 [==============================] - 34s 18ms/step - loss: 0.0224 - accuracy: 0.9930
Epoch 5/5
313/313 [==============================] - 2s 5ms/step - loss: 0.0538 - accuracy: 0.9833
Test accuracy of CNN model: 0.983299970626831


Activity 5:

Test accuracy of baseline model: 0.9757000207901001

Test accuracy of CNN model: 0.983299970626831

CNNs are typically used for image-related tasks and often outperform feedforward networks for image classification tasks like MNIST.